pkg/engine/lifeycletest/pulumi_test.go

```diff
@@ -942,6 +942,159 @@ func TestSingleResourceIgnoreChanges(t *testing.T) {
 	}, []string{"a", "b"}, []deploy.StepOp{deploy.OpUpdate})
 }
 
+func objectDiffToDetailedDiff(prefix string, d *resource.ObjectDiff) map[string]plugin.PropertyDiff {
+	ret := map[string]plugin.PropertyDiff{}
+	for k, vd := range d.Updates {
+		var nestedPrefix string
+		if prefix == "" {
+			nestedPrefix = string(k)
+		}
+		if prefix != "" {
+			nestedPrefix = fmt.Sprintf("%s.%s", prefix, string(k))
+		}
+		for kk, pd := range valueDiffToDetailedDiff(nestedPrefix, vd) {
+			ret[kk] = pd
+		}
+	}
+	return ret
+}
+
+func arrayDiffToDetailedDiff(prefix string, d *resource.ArrayDiff) map[string]plugin.PropertyDiff {
+	ret := map[string]plugin.PropertyDiff{}
+	for i, vd := range d.Updates {
+		for kk, pd := range valueDiffToDetailedDiff(fmt.Sprintf("%s[%d]", prefix, i), vd) {
+			ret[kk] = pd
+		}
+	}
+	return ret
+}
+
+func valueDiffToDetailedDiff(prefix string, vd resource.ValueDiff) map[string]plugin.PropertyDiff {
+	ret := map[string]plugin.PropertyDiff{}
+	if vd.Object != nil {
+		for kk, pd := range objectDiffToDetailedDiff(prefix, vd.Object) {
+			ret[kk] = pd
+		}
+	} else if vd.Array != nil {
+		for kk, pd := range arrayDiffToDetailedDiff(prefix, vd.Array) {
+			ret[kk] = pd
+		}
+	} else {
+		ret[prefix] = plugin.PropertyDiff{Kind: plugin.DiffUpdate}
+	}
+	return ret
+}
+
+func TestReplaceOnChanges(t *testing.T) {
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				DiffF: func(urn resource.URN, id resource.ID,
```

Comment:
It seems like we should have something built-in to produce a `plugin.DiffResult` from two `resource.PropertyMaps` (and potentially a set of replace keys).  I was surprised I had to implement all this myself.  Did I miss where this is defined?  cc @pgavlin

Reply:
It looks like there is code like this in most of our native providers: https://github.com/pulumi/pulumi-azure-native/blob/ae7f380268c97b951a76848ec45b3bafa22d4338/provider/pkg/provider/diff.go.  It's mostly the same sort of thing implemented here - but with some slightly special-cased to pull the "set of replace keys" from the specific provider schema.  It feels like that implementation could/should be pulled into the core and generalized so it doesn't have to be reimplemented in every provider (and could also be reused in tests here).